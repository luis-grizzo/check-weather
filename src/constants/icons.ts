import clear_day from '@public/clear_day.svg'
import clear_night from '@public/clear_night.svg'
import partly_cloudy_day from '@public/partly_cloudy_day.svg'
import partly_cloudy_night from '@public/partly_cloudy_night.svg'
import rainy from '@public/rainy.svg'
import thunderstorm from '@public/thunderstorm.svg'
import weather_snowy from '@public/weather_snowy.svg'
import mist from '@public/mist.svg'
import weather_mix from '@public/weather_mix.svg'
import foggy from '@public/foggy.svg'
import airwave from '@public/airwave.svg'
import storm from '@public/storm.svg'
import tornado from '@public/tornado.svg'

import refresh from '@public/refresh.svg'
import cloud_off from '@public/cloud_off.svg'
import air from '@public/air.svg'
import visibility from '@public/visibility.svg'
import water_drop from '@public/water_drop.svg'
import info from '@public/info.svg'
import error from '@public/error.svg'
import progress_activity from '@public/progress_activity.svg'

export const weatherIcons = {
  clear_day: {
    src: clear_day,
    alt: 'Clear day icon'
  },
  clear_night: {
    src: clear_night,
    alt: 'Clear night icon'
  },
  partly_cloudy_day: {
    src: partly_cloudy_day,
    alt: 'Partly cloudy day icon'
  },
  partly_cloudy_night: {
    src: partly_cloudy_night,
    alt: 'Partly cloudy night icon'
  },
  rainy: {
    src: rainy,
    alt: 'Rainy icon'
  },
  thunderstorm: {
    src: thunderstorm,
    alt: 'Thunderstorm icon'
  },
  weather_snowy: {
    src: weather_snowy,
    alt: 'Weather snowy icon'
  },
  mist: {
    src: mist,
    alt: 'Mist icon'
  },
  weather_mix: {
    src: weather_mix,
    alt: 'Weather mix icon'
  },
  foggy: {
    src: foggy,
    alt: 'Foggy icon'
  },
  airwave: {
    src: airwave,
    alt: 'Airwave icon'
  },
  storm: {
    src: storm,
    alt: 'Storm icon'
  },
  tornado: {
    src: tornado,
    alt: 'Tornado icon'
  }
}

export const layoutIcons = {
  refresh,
  cloud_off,
  air,
  visibility,
  water_drop,
  info,
  error,
  progress_activity
}

export const weatherIconsArray = Object.values(weatherIcons)
