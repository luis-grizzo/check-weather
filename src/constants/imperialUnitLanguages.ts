export const imperialUnitLanguages = ['en-US', 'es-US']
