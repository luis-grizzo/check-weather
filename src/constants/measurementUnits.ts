export const imperialUnitLanguages = ['en-US', 'es-US']

export const validMeasurementUnits = ['imperial', 'metric']
