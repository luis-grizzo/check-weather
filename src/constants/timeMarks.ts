export const hourInMilliseconds = 3_600_000

export const hourInSeconds = 3_600
