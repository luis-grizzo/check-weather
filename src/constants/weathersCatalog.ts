import { weatherIcons } from './icons'

export const weathersCatalog = {
  clear: {
    color: 'before:from-sky-300',
    icon: {
      day: weatherIcons.clear_day,
      night: weatherIcons.clear_night
    }
  },
  clouds: {
    color: 'before:from-blue-300',
    icon: {
      day: weatherIcons.partly_cloudy_day,
      night: weatherIcons.partly_cloudy_night
    }
  },
  drizzle: {
    color: 'before:from-indigo-300',
    icon: {
      day: weatherIcons.rainy,
      night: weatherIcons.rainy
    }
  },
  rain: {
    color: 'before:from-violet-300',
    icon: {
      day: weatherIcons.rainy,
      night: weatherIcons.rainy
    }
  },
  thunderstorm: {
    color: 'before:from-purple-300',
    icon: {
      day: weatherIcons.thunderstorm,
      night: weatherIcons.thunderstorm
    }
  },
  snow: {
    color: 'before:from-teal-300',
    icon: {
      day: weatherIcons.weather_snowy,
      night: weatherIcons.weather_snowy
    }
  },
  mist: {
    color: 'before:from-slate-300',
    icon: {
      day: weatherIcons.mist,
      night: weatherIcons.mist
    }
  },
  haze: {
    color: 'before:from-slate-300',
    icon: {
      day: weatherIcons.weather_mix,
      night: weatherIcons.weather_mix
    }
  },
  fog: {
    color: 'before:from-slate-300',
    icon: {
      day: weatherIcons.foggy,
      night: weatherIcons.foggy
    }
  },
  smoke: {
    color: 'before:from-slate-300',
    icon: {
      day: weatherIcons.foggy,
      night: weatherIcons.foggy
    }
  },
  dust: {
    color: 'before:from-yellow-300',
    icon: {
      day: weatherIcons.airwave,
      night: weatherIcons.airwave
    }
  },
  sand: {
    color: 'before:from-orange-300',
    icon: {
      day: weatherIcons.airwave,
      night: weatherIcons.airwave
    }
  },
  ash: {
    color: 'before:from-red-300',
    icon: {
      day: weatherIcons.mist,
      night: weatherIcons.mist
    }
  },
  squall: {
    color: 'before:from-pink-300',
    icon: {
      day: weatherIcons.storm,
      night: weatherIcons.storm
    }
  },
  tornado: {
    color: 'before:from-pink-300',
    icon: {
      day: weatherIcons.tornado,
      night: weatherIcons.tornado
    }
  }
}
