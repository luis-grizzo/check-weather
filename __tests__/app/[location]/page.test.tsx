import { render } from '@testing-library/react'

import Location from '@/app/[location]/page'
import { hourInMilliseconds } from '@/constants/timeMarks'

const unixTimestamp = 1718908012
const secondInMilliseconds = 1_000

jest.useFakeTimers()
jest.setSystemTime(
  new Date(unixTimestamp * 1_000 + hourInMilliseconds - secondInMilliseconds)
)

jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      refresh: jest.fn()
    }
  }
}))

global.fetch = jest.fn().mockImplementation(() => ({
  ok: true,
  json: jest.fn().mockImplementation(() => ({
    coord: {
      lon: -48.6,
      lat: -22.3
    },
    weather: [
      {
        id: 800,
        main: 'Clear',
        description: 'clear sky',
        icon: '01d'
      }
    ],
    base: 'stations',
    main: {
      temp: 30.21,
      feels_like: 28.33,
      temp_min: 30.21,
      temp_max: 30.27,
      pressure: 1015,
      humidity: 17,
      sea_level: 1015,
      grnd_level: 952
    },
    visibility: 10000,
    wind: {
      speed: 0.83,
      deg: 43,
      gust: 1.91
    },
    clouds: {
      all: 0
    },
    dt: unixTimestamp,
    sys: {
      type: 2,
      id: 53165,
      country: 'BR',
      sunrise: 1718877183,
      sunset: 1718915943
    },
    timezone: -10800,
    id: 3460005,
    name: 'JaÃº',
    cod: 200
  }))
}))

describe('Location', () => {
  afterAll(() => {
    jest.clearAllMocks()
    jest.clearAllTimers()
  })

  it('should match snapshot', async () => {
    const encodedParams = encodeURIComponent('-22.2,-48.5,metric')

    const Resolved = await Location({
      params: { location: encodedParams }
    })

    const { container } = render(Resolved)

    expect(container).toMatchSnapshot()
  })

  it('should throw if params are wrong', async () => {
    const encodedParams = encodeURIComponent('NAN,NAN,notValid')

    await expect(
      async () => await Location({ params: { location: encodedParams } })
    ).rejects.toThrow()
  })
})
