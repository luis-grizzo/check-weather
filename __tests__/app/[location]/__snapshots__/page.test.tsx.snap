// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location should match snapshot 1`] = `
<div>
  <header
    class="flex items-center justify-between container mx-auto px-8 py-6"
  >
    <div
      class="flex flex-col gap-2"
    >
      <span
        class="text-lg"
      >
        Jaú, Brazil
      </span>
      <span
        class="text-xs"
      >
        3:26 PM - Thursday, June 20, 2024
      </span>
    </div>
    <div
      class="flex flex-col items-center gap-2"
    >
      <button
        class="flex items-center justify-center p-2 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-full disabled:opacity-50 disabled:cursor-not-allowed transition-opacity"
        disabled=""
        role="button"
      >
        <img
          alt="Loading icon"
          class="w-6 aspect-square animate-spin"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
  </header>
  <main
    class="relative flex flex-col items-center justify-center gap-8 h-full container mx-auto px-8 py-6 before:absolute before:w-3/4 before:max-w-96 before:aspect-square before:bg-[radial-gradient(circle_at_center,_var(--tw-gradient-stops))] before:from-blue-300 before:to-transparent before:to-70%  before:rounded-full before:animate-pulse before:-z-50"
  >
    <div
      class="flex items-center gap-4"
    >
      <img
        alt="Clear day icon"
        class="w-24 aspect-square"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <span
        class="text-8xl"
      >
        30°
      </span>
    </div>
    <div
      class="flex flex-wrap justify-center gap-2"
    >
      <span
        class="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur"
      >
        Feels like 28°
      </span>
      <span
        class="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur"
      >
        Min 30°
      </span>
      <span
        class="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur"
      >
        Max 30°
      </span>
      <span
        class="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur"
      >
        Sunrise 6:53 AM
      </span>
      <span
        class="flex items-center gap-2 px-2 py-1 text-xs bg-neutral-100/60 border-1 border-neutral-600/10 text-neutral-950 rounded-lg backdrop-blur"
      >
        Sunset 5:39 PM
      </span>
    </div>
    <span
      class="text-base capitalize"
    >
      clear sky
    </span>
    <div
      class="flex items-center justify-between gap-8 p-8 bg-neutral-100/60 border-1 border-neutral-600/10 rounded-2xl backdrop-blur"
    >
      <div
        class="flex flex-col items-center gap-2"
      >
        <img
          alt="Wind speed icon"
          class="w-8 aspect-square"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <span
          class="text-lg text-nowrap"
        >
          1 m/s
        </span>
        <span
          class="text-xs"
        >
          Wind
        </span>
      </div>
      <div
        class="flex flex-col items-center gap-2"
      >
        <img
          alt="Humidity icon"
          class="w-8 aspect-square"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <span
          class="text-lg text-nowrap"
        >
          17%
        </span>
        <span
          class="text-xs"
        >
          Humidity
        </span>
      </div>
      <div
        class="flex flex-col items-center gap-2"
      >
        <img
          alt="Visibility icon"
          class="w-8 aspect-square"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
        <span
          class="text-lg text-nowrap"
        >
          10 km
        </span>
        <span
          class="text-xs"
        >
          Visibility
        </span>
      </div>
    </div>
  </main>
</div>
`;
